### PR TITLE
Fix `Inject` instance for lists

### LINKED ISSUE #1254: Error injecting and interpreting lists
Hi, i've observed that `Dhall.inject` and `Dhall.input Dhall.auto` doesn't behave correctly with lists. In a ghci session with 0b3dae7
```
> pretty $ (Dhall.embed Dhall.inject) (["x"] :: [Text])
"[ \"x\" ] : Text"
> pretty $ (Dhall.embed Dhall.inject) ([1] :: [Int])
"[ +1 ] : Integer"
> pretty $ (Dhall.embed Dhall.inject) ([[1]] :: [[Int]])
"[ [ +1 ] : Integer ] : List Integer"
> Dhall.input Dhall.auto $ "[ +1 ]"
*** Exception:
 Error [0m: Expression doesn't match annotation

- Integer
+ List …

1│[ +1 ] : Integer

(input):1:1
> Dhall.input (Dhall.list Dhall.integer) $ "[ +1 ]" -- using explicit types works
[1]
```
It should be:
```
> pretty $ (Dhall.embed Dhall.inject) (["x"] :: [Text])
"[ \"x\" ] : List Text"
> pretty $ (Dhall.embed Dhall.inject) ([1] :: [Int])
"[ +1 ] : List Integer"
> pretty $ (Dhall.embed Dhall.inject) ([[1]] :: [[Int]])
"[ [ +1 ] : List Integer ] : List (List Integer)"
> Dhall.input Dhall.auto $ "[ +1 ]"
[1]
```

### PR DESCRIPTION
**Pull request by Gabriel439**
Fixes #1254

This also adds tests for conversions back and forth between Dhall and
Haskell values

### CODE REVIEW

File: dhall/tests/Dhall/Test/Dhall.hs [hunk 1]
```diff
@@ -239,3 +246,64 @@ shouldHandleUnionsCorrectly =
...
+shouldConvertHaskellToDhallCorrectly :: TestTree
+shouldConvertHaskellToDhallCorrectly =
+    testGroup
+        "Marshall Haskell to Dhall code"
+        [ "True" `correspondsTo` True
+        , "False" `correspondsTo` False
+        , "2" `correspondsTo` (2 :: Natural)
+        , "+2" `correspondsTo` (2 :: Integer)
+        , "2.0" `correspondsTo` (2.0 :: Double)
+--      , "2.0" `correspondsTo` (2.0 :: Scientific)
```
Comment (sjakobi) [inline dhall/tests/Dhall/Test/Dhall.hs]: Why is this commented out?

File: dhall/tests/Dhall/Test/Dhall.hs [hunk 2]
```diff
@@ -239,3 +246,64 @@ shouldHandleUnionsCorrectly =
...
+        , "\"ABC\"" `correspondsTo` ("ABC" :: Data.Text.Text)
+        , "\"ABC\"" `correspondsTo` ("ABC" :: Data.Text.Lazy.Text)
+--      , "\"ABC\"" `correspondsTo` ("ABC" :: String)
```
Comment (sjakobi) [inline dhall/tests/Dhall/Test/Dhall.hs]: Dito.

File: dhall/tests/Dhall/Test/Dhall.hs (hunk 1 above)
Comment (Gabriel439) [reply dhall/tests/Dhall/Test/Dhall.hs]: Missing instance, which I can add for symmetry

Comment (sjakobi) [reply dhall/tests/Dhall/Test/Dhall.hs]: Can you add a comment regarding the missing instance?

File: dhall/tests/Dhall/Test/Dhall.hs (hunk 2 above)
Comment (sjakobi) [reply dhall/tests/Dhall/Test/Dhall.hs]: Same here.

**Comment by Gabriel439** [comment]
Actually, I realized that it was easier to just add the missing instances, so I'll do that

**Comment by jneira** [comment]
Thanks for the quick fix!

**Comment by Gabriel439** [comment]
@jneira: You're welcome! 🙂

**Pull request merged**